core/api.go

```diff
@@ -250,15 +252,21 @@ func (a *API) authnHandler(handler http.Handler) http.Handler {
 		req, err := auth.Authenticate(req)
 		if err != nil {
 			errorFormatter.Write(req.Context(), rw, errNotAuthenticated)
+			return
 		}
 		handler.ServeHTTP(rw, req)
 	})
 }
 
-func authzHandler(handler http.Handler) http.Handler {
+func (a *API) authzHandler(handler http.Handler) http.Handler {
+	auth := authz.NewAuthorizer(a.raftDB, grantPrefix, policyByRoute)
 	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
-		if !authz.Authorized(req.Context()) {
-			errorFormatter.Write(req.Context(), rw, errNotAuthorized)
+		err := auth.Authorize(req)
+		if err != nil {
+			// TODO(tessr/dominic): Switch the following to:
+			// errorFormatter.Write(req.Context(), rw, err)
+			// once dashboard can handle ErrNotAuthorized as well as errNotAuthenticated
+			errorFormatter.Write(req.Context(), rw, errNotAuthenticated)
```

Comment:
I think it's still necessary, because dashboard doesn't know how to interpret errNotAuthorized, but this is better.